src/imputation/impute_civ_def.py

```diff
@@ -0,0 +1,259 @@
+    num_valid = df.groupby(class_name)["valid_civdef_val"].transform(sum)
+
+    # exclude classes that are not valid
+    valid_class_cond = ~(df[class_name].str.contains("nan"))
+
+    # create new bool column to flag empty classes
+    df[new_col_name] = valid_class_cond &  (num_valid == 0) 
+
+    df = df.drop("valid_civdef_val", axis=1)
+
+    return df
+
+
+def prep_cd_imp_classes(df: pd.DataFrame) -> pd.DataFrame:
+    """Create new columns to use for imputation classes,
+
+    Also create boolean columns to flag for when these contain no
+    valid entries that can be used for imputation.
+    """
+    # create imputation classes based on product group and rusic
+    df = tmi.create_imp_class_col(df, "201", "rusic", "pg_sic_class")
+
+    # flag empty pg_sic_classes in new bool col "empty_pgsic_class"
+    df = calc_empty_group(df, "pg_sic_class", "empty_pgsic_group")
+
+    # create a new imputation class called "pg_class" based on product group
+    # (col 201) only, to be used when imputation cannot be performed using 
+    # both prodcut group and SIC.
+    df = tmi.create_imp_class_col(df, "201", None, "pg_class")
+
+    # flag empty pg_classes in new bool col "empty_pg_class"
+    df = calc_empty_group(df, "pg_class", "empty_pg_group")
+
+    return df
+
+
+def random_assign_civdef(
+    df: pd.DataFrame, 
+    proportions: Tuple[float, float]
+) -> pd.DataFrame:
+    """Assign "C" for civil or "D" for defence randomly based on
+        the proportions supplied.
+    """
+    np.random.seed(seed=42)
+    df["200_imputed"] = np.random.choice(
+        ["C", "D"], 
+        size = len(df), 
+        p = proportions
+    )
+    return df
+
+
+def apply_civdev_imputation(
+    df: pd.DataFrame, 
+    pgsic_dict: Dict[str, Tuple[float, float]],
+    pg_dict: Dict[str, Tuple[float, float]]
+) -> pd.DataFrame:
+    """Apply imputation for R&D type for non-responders and 'No R&D'.
+
+    Values in column 200 (R&D type) are imputed with either "C" for civl or
+    "D" for defence, based on ratios in the same imputation class in 
+    clear responders.
+
+    This is done in three passes for successively smaller imputation classes:
+    - The first class consists of all clear responders
+    - The second set of imputation classes are based on product group only
+    - The final set of imputation classes are bassed on product group and SIC
+    
+    Args:
+        df (pd.DataFrame): The dataframe of all responses
+        pgsic_dict (Dict[str, Tuple(float, float)]): Dictionary with proportions
+            to use in imputation based on product group and SIC.
+        pg_dict (Dict[str, Tuple(float, float)]): Dictionary with proportions
+            to use in imputation based on product group only.
+
+    Returns:
+        pd.DataFrame: The same dataframe with R&D type imputed.
+    '"""
+    df["200_original"] = df["200"]
+    df["200_imputed"] = df["200"]
+    df["200_imp_marker"] = "N/A"
```

Comment:
Should this be 'nan' or `np.nan` for consistency maybe?

Reply:
good question! Actually, following what we have in TMI step 3, I actually want the string "N/A" as in "not applicable". It comes out as the string "N/A" in excel, but if it's ambiguous because it could be interpreted as a null in some contexts, we could possibly change it to "no_imputation"

Reply:
Yeh I think 'no_imputation' may be a bit clearer - I don't know if pandas might interpret these as `nan` in some scenarios.

Reply:
added this, also in tmi_imputation